File: leetcode/0096. Unique Binary Search Trees/Solution.swift
// https://leetcode.com/problems/unique-binary-search-trees/

class Solution {
    func numTrees(_ n: Int) -> Int {
        if n <= 1 {
            return 1
        }
        
        var f = [Int](repeating: 0, count: n + 1)
        f[0] = 1
        f[1] = 1
        for r in 2...n {
            for k in 1...r {
                f[r] += f[k - 1] * f[r - k]
            }
        }        
        
        return f[n]
    }
}

File: leetcode/0096. Unique Binary Search Trees/Solution.py
# https://leetcode.com/problems/unique-binary-search-trees/

class Solution:
    def numTrees(self, n: int) -> int:
        if n <= 1:
            return 1
        
        f = [0] * (n + 1)
        f[0] = 1
        f[1] = 1
        for r in range(2, n + 1):
            for k in range(1, r + 1):
                f[r] += f[k - 1] * f[r - k]

        return f[n]

s = Solution()
assert s.numTrees(3) == 5